sdk/identity/azure-identity/src/client_certificate_credential.cpp

```diff
@@ -69,12 +159,91 @@ template <> struct UniqueHandleHelper<EVP_MD_CTX>
 
 template <typename T>
 using UniqueHandle = Azure::Core::_internal::UniqueHandle<T, UniqueHandleHelper>;
+
+std::tuple<CertificateThumbprint, UniquePrivateKey> ReadCertificate(const std::string& path)
+{
+  // Open certificate file, then get private key and X509:
+  UniqueHandle<BIO> bio(BIO_new_file(path.c_str(), "r"));
+  if (!bio)
+  {
+    throw AuthenticationException("Failed to open certificate file.");
+  }
+
+  UniquePrivateKey pkey{PEM_read_bio_PrivateKey(bio.get(), nullptr, nullptr, nullptr)};
+  if (!pkey)
+  {
+    throw AuthenticationException("Failed to read certificate private key.");
+  }
+
+  UniqueHandle<X509> x509{PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr)};
+  if (!x509)
+  {
+    std::ignore = BIO_seek(bio.get(), 0);
+    x509.reset(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr));
+    if (!x509)
+    {
+      throw AuthenticationException("Failed to read X509 section.");
+    }
+  }
+
+  CertificateThumbprint thumbprint(EVP_MAX_MD_SIZE);
+  // Get certificate thumbprint:
+  unsigned int mdLen = 0;
+  const auto digestResult = X509_digest(x509.get(), EVP_sha1(), thumbprint.data(), &mdLen);
+
+  if (!digestResult)
+  {
+    throw AuthenticationException("Failed to get certificate thumbprint.");
+  }
+
+  // Drop unused buffer space:
+  const auto mdLenSz = static_cast<decltype(thumbprint)::size_type>(mdLen);
+  if (thumbprint.size() > mdLenSz)
+  {
+    thumbprint.resize(mdLenSz);
+  }
+
+  return std::make_tuple(thumbprint, std::move(pkey));
+}
+
+std::vector<unsigned char> SignPkcs1Sha256(PrivateKey key, const uint8_t* data, size_t size)
+{
+  UniqueHandle<EVP_MD_CTX> mdCtx(EVP_MD_CTX_new());
+  if (!mdCtx)
+  {
+    return {};
+  }
+  EVP_PKEY_CTX* signCtx = nullptr;
+  if ((EVP_DigestSignInit(mdCtx.get(), &signCtx, EVP_sha256(), nullptr, static_cast<EVP_PKEY*>(key))
+       == 1)
+      && (EVP_PKEY_CTX_set_rsa_padding(signCtx, RSA_PKCS1_PADDING) == 1))
+  {
+    size_t sigLen = 0;
+    if (EVP_DigestSign(mdCtx.get(), nullptr, &sigLen, nullptr, 0) == 1)
+    {
+      std::vector<unsigned char> sigVec(sigLen);
+      if (EVP_DigestSign(mdCtx.get(), sigVec.data(), &sigLen, data, size) == 1)
+      {
+        return sigVec;
+      }
+    }
+  }
+  return {};
+}
+#endif
 } // namespace
 
+#ifdef WIN32
+void Azure::Identity::_detail::FreeNcryptKeyImpl(void* pkey)
+{
+  NCryptFreeObject(reinterpret_cast<NCRYPT_KEY_HANDLE>(pkey));
```

Comment:
Not 100% sure but I think you might be able to use `static_cast` here instead of reinterpret_cast (because the source is a void *).

Reply:
I can't; I'm getting `invalid type conversion C/C++(171)`.